### PR TITLE
fix(core)!: `$(...).[last,first,all]()` defaulting to the first output instead of the output that connects the nodes

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

This PR makes `$(...).[last|first|all]()` default to the the output the node was connected to.

Before it would default to the first output.

<details><summary>Example Workflow</summary>
<p>

```json
{
  "meta": {
    "templateCredsSetupCompleted": true,
    "instanceId": "6944c3a4f9d5af1d486eb3bd76ceeae78d947e96a0d9e6c404bb1bcec6a36054"
  },
  "nodes": [
    {
      "parameters": {
        "path": "e9586c98-e50e-43fb-9c6b-2ff1deaaa0c9",
        "options": {}
      },
      "id": "669a68fd-cc30-45ff-88f4-6a5455bcf66a",
      "name": "Webhook1",
      "type": "n8n-nodes-base.webhook",
      "typeVersion": 2,
      "position": [
        180,
        700
      ],
      "webhookId": "e9586c98-e50e-43fb-9c6b-2ff1deaaa0c9"
    },
    {
      "parameters": {
        "conditions": {
          "options": {
            "caseSensitive": true,
            "leftValue": "",
            "typeValidation": "strict"
          },
          "conditions": [
            {
              "id": "1fff886f-3d13-4fbf-b0fb-7e2f845937c0",
              "leftValue": "={{ true }}",
              "rightValue": "",
              "operator": {
                "type": "boolean",
                "operation": "true",
                "singleValue": true
              }
            }
          ],
          "combinator": "and"
        },
        "options": {}
      },
      "id": "588e25bd-940c-498b-91e3-d768f45cc616",
      "name": "true",
      "type": "n8n-nodes-base.if",
      "typeVersion": 2,
      "position": [
        460,
        460
      ]
    },
    {
      "parameters": {
        "conditions": {
          "options": {
            "caseSensitive": true,
            "leftValue": "",
            "typeValidation": "strict"
          },
          "conditions": [
            {
              "id": "1fff886f-3d13-4fbf-b0fb-7e2f845937c0",
              "leftValue": "={{ false }}",
              "rightValue": "",
              "operator": {
                "type": "boolean",
                "operation": "true",
                "singleValue": true
              }
            }
          ],
          "combinator": "and"
        },
        "options": {}
      },
      "id": "dddc1d50-495c-403e-a17a-5305a3b277a8",
      "name": "false",
      "type": "n8n-nodes-base.if",
      "typeVersion": 2,
      "position": [
        460,
        660
      ]
    },
    {
      "parameters": {
        "conditions": {
          "options": {
            "caseSensitive": true,
            "leftValue": "",
            "typeValidation": "strict"
          },
          "conditions": [
            {
              "id": "1fff886f-3d13-4fbf-b0fb-7e2f845937c0",
              "leftValue": "={{ true }}",
              "rightValue": "",
              "operator": {
                "type": "boolean",
                "operation": "true",
                "singleValue": true
              }
            }
          ],
          "combinator": "and"
        },
        "options": {}
      },
      "id": "b794c9be-7d16-47fb-9a5d-d3816c44e247",
      "name": "false1",
      "type": "n8n-nodes-base.if",
      "typeVersion": 2,
      "position": [
        460,
        860
      ]
    },
    {
      "parameters": {
        "assignments": {
          "assignments": [
            {
              "id": "41531ffa-e84c-4843-a401-45b93d520f97",
              "name": "test",
              "value": "={{ $('false').last().json.code }}",
              "type": "string"
            }
          ]
        },
        "options": {}
      },
      "id": "56928338-14d7-4e70-8979-d091fc8da23e",
      "name": "Edit fields 2",
      "type": "n8n-nodes-base.set",
      "typeVersion": 3.3,
      "position": [
        840,
        660
      ]
    },
    {
      "parameters": {
        "assignments": {
          "assignments": [
            {
              "id": "41531ffa-e84c-4843-a401-45b93d520f97",
              "name": "test",
              "value": "={{ $('true').last().json.code }}",
              "type": "string"
            }
          ]
        },
        "options": {}
      },
      "id": "c71938b7-fe93-4a6a-8c90-10ef02f1d56f",
      "name": "Edit fields 1",
      "type": "n8n-nodes-base.set",
      "typeVersion": 3.3,
      "position": [
        840,
        460
      ]
    },
    {
      "parameters": {
        "assignments": {
          "assignments": [
            {
              "id": "41531ffa-e84c-4843-a401-45b93d520f97",
              "name": "test",
              "value": "={{ $('false1').last().json.code }}",
              "type": "string"
            }
          ]
        },
        "options": {}
      },
      "id": "aab4f1c7-465b-474b-b769-4afd10232f1a",
      "name": "Edit fields 3",
      "type": "n8n-nodes-base.set",
      "typeVersion": 3.3,
      "position": [
        840,
        860
      ]
    }
  ],
  "connections": {
    "Webhook1": {
      "main": [
        [
          {
            "node": "true",
            "type": "main",
            "index": 0
          },
          {
            "node": "false",
            "type": "main",
            "index": 0
          },
          {
            "node": "false1",
            "type": "main",
            "index": 0
          },
          {
            "node": "Edit fields 3",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "true": {
      "main": [
        [
          {
            "node": "Edit fields 1",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "false": {
      "main": [
        null,
        [
          {
            "node": "Edit fields 2",
            "type": "main",
            "index": 0
          }
        ]
      ]
    },
    "false1": {
      "main": [
        null,
        [
          {
            "node": "Edit fields 3",
            "type": "main",
            "index": 0
          }
        ]
      ]
    }
  },
  "pinData": {
    "Webhook1": [
      {
        "name": "First item",
        "code": 1
      },
      {
        "name": "Second item",
        "code": 2
      }
    ]
  }
}
```

</p>
</details> 

Before it would always evaluate output 0 (true branch) 
image

Now it will take the output that connects the nodes. If the user wants another output they'd have to specify if like `$(...).last(1)`
image

Workflows that rely on something like the third example will break with this change.
Workflows that rely on `last`, `first` and `all` always returning the first output will also break.

The ticket is marked as a bug, meaning this was never the intended behavior. I marked this as a breaking change regardless. 

We'd also want to update the docs to make it clear how those functions work: https://docs.n8n.io/code/builtin/output-other-nodes/

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/PAY-1400/bug-dollarlast-uses-wrong-branch
https://github.com/n8n-io/n8n-docs/pull/2167

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included. 
- [ ] ~PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)~

### CODE REVIEW

**Review by krynble — DISMISSED** [review]
Thanks for addressing this bug and adding the tests! Great work!

**Review by ivov — CHANGES_REQUESTED** [review]
Thank you for fixing this! If this is a breaking change, let's please update the [breaking changes doc](https://github.com/n8n-io/n8n/blob/master/packages/cli/BREAKING-CHANGES.md) as well. Else let's remove `!` from the title.

**Comment by ivov** [comment]
Follow-up: https://linear.app/n8n/issue/PAY-1692

File: packages/cli/BREAKING-CHANGES.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
 
 This list shows all the versions which include breaking changes and how to upgrade.
 
+## 1.46.0
```
Comment (ivov) [inline packages/cli/BREAKING-CHANGES.md:5]: Next version will be 1.47.0

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.47.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.47.0)

**Comment by janober** [comment]
Got released with [`n8n@1.47.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.47.0)

**Pull request merged**